File: tests/fixtures/api-baseline/lib/problem/local_error.go
package problem

import "context"

const (
	localErrorID      = "local-error"
	localErrorTitle   = "Local Error"
	localErrorVersion = "v1"
)

// ErrorSink receives unexpected problems captured by [LocalError]. In
// production this forwards to the telemetry path; tests inject a recording sink.
type ErrorSink interface {
	// Capture reports problem, returning any delivery error.
	Capture(ctx context.Context, problem Problem) error
}

// NoopErrorSink is an [ErrorSink] that discards every capture.
type NoopErrorSink struct{}

// Capture discards problem and always succeeds.
func (NoopErrorSink) Capture(_ context.Context, _ Problem) error {
	return nil
}

// LocalError wraps unexpected errors into a `local-error` [Problem] carrying the
// message and stack in `data`, and captures it on a [ErrorSink]. The type URI
// flows through the single-source builder so even local errors share one
// identity shape.
type LocalError struct {
	// Sink is the destination for captured local-error problems.
	Sink ErrorSink
	// Portal builds the local-error type URI.
	Portal ErrorPortal
}

// NewLocalError creates a wrapper writing captures to sink, using
// [LocalErrorPortal] for its type URIs.
func NewLocalError(sink ErrorSink) LocalError {
	return LocalError{Sink: sink, Portal: LocalErrorPortal()}
}

// NewLocalErrorWithPortal creates a wrapper writing captures to sink and
// building type URIs from portal.
func NewLocalErrorWithPortal(sink ErrorSink, portal ErrorPortal) LocalError {
	return LocalError{Sink: sink, Portal: portal}
}

// Wrap folds cause and stack into a `local-error` [Problem], captures it on the
// sink, and returns the envelope. A capture failure is returned alongside the
// (still valid) envelope so callers never lose the problem.
func (local LocalError) Wrap(ctx context.Context, cause error, stack string) (Problem, error) {
	message := ""
	if cause != nil {
		message = cause.Error()
	}
	typeURI, err := TypeURI(local.Portal, localErrorVersion, localErrorID)
	if err != nil {
		typeURI = blankType
	}
	envelope := Problem{
		Type:        typeURI,
		Title:       localErrorTitle,
		Status:      defaultProblemStatus,
		Detail:      message,
		Recoverable: false,
		Data: map[string]any{
			"message": message,
			"stack":   stack,
		},
	}
	if captureErr := local.Sink.Capture(ctx, envelope); captureErr != nil {
		return envelope, captureErr
	}
	return envelope, nil
}

File: packages/diene_problems/lib/src/local_error.dart
/// LocalError — wraps unexpected client-side exceptions into a Problem
/// (goal: dart frontends wrap unexpected exceptions into a `LocalError` Problem,
/// message + stacktrace in `data`, rendered by a Problem visualizer shipped in
/// flutter-base).
///
/// The type URI flows through the single-source builder ([problemTypeUri]) so
/// even local errors share one identity shape; the `data` payload carries the
/// message and stack trace the visualizer surfaces.
library;

import 'problem.dart';
import 'problem_type_uri.dart';

/// Receives unexpected problems captured by [LocalError].
///
/// In production this forwards to the frontend telemetry/Faro path (Faro owns
/// runtime emission; this seam stays owned by the interfaces/telemetry layer —
/// see node note). Tests inject a recording sink.
abstract interface class ErrorSink {
  /// Captures [problem] for reporting.
  Future<void> capture(Problem problem);
}

/// An [ErrorSink] that discards every capture.
final class NoopErrorSink implements ErrorSink {
  /// Creates a no-op sink.
  const NoopErrorSink();

  @override
  Future<void> capture(Problem problem) async {}
}

/// Wraps unexpected exceptions into a `LocalError` Problem.
///
/// ```dart
/// final Problem problem = await LocalError(sink).wrap(error, stack);
/// ```
final class LocalError {
  /// Creates a local-error wrapper writing captures to [sink].
  const LocalError(this.sink, {this.portal = ErrorPortal.localError});

  /// Destination for captured local-error problems.
  final ErrorSink sink;

  /// Portal used to build the local-error type URI.
  final ErrorPortal portal;

  /// Wraps [error] + [stackTrace] into a `LocalError` Problem, captures it on
  /// [sink], and returns it.
  Future<Problem> wrap(Object error, StackTrace stackTrace) async {
    final String message = error.toString();
    final Problem problem = Problem(
      type: problemTypeUri(portal: portal, version: 'v1', id: 'local_error'),
      title: 'Local Error',
      status: 500,
      detail: message,
      recoverable: false,
      data: <String, Object?>{
        'message': message,
        'stackTrace': stackTrace.toString(),
      },
    );
    await sink.capture(problem);
    return problem;
  }
}
